File: movie_randomizer.go
// movie_randomizer.go
package main

import (
	"bufio"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

type Movie struct {
	ID        int    `json:"id"`
	Title     string `json:"title"`
	Genre     string `json:"genre"`
	Director  string `json:"director"`
	Year      int    `json:"year"`
	Watched   bool   `json:"watched"`
	Rating    int    `json:"rating"`
	Notes     string `json:"notes"`
	AddedDate string `json:"added_date"`
}

type MoviesData struct {
	Movies []Movie `json:"movies"`
}

type MovieRandomizer struct {
	movies  []Movie
	nextID  int
}

func NewMovieRandomizer() *MovieRandomizer {
	rand.Seed(time.Now().UnixNano())
	return &MovieRandomizer{
		movies:  []Movie{},
		nextID:  1,
	}
}

func (r *MovieRandomizer) AddMovie(title, genre, director string, year int, watched bool, rating int, notes string) (Movie, error) {
	if rating < 1 || rating > 10 {
		return Movie{}, fmt.Errorf("оценка должна быть от 1 до 10")
	}
	currentYear := time.Now().Year()
	if year < 1900 || year > currentYear {
		return Movie{}, fmt.Errorf("год должен быть от 1900 до %d", currentYear)
	}
	if title == "" || genre == "" {
		return Movie{}, fmt.Errorf("название и жанр не могут быть пустыми")
	}
	if director == "" {
		director = "Неизвестен"
	}
	movie := Movie{
		ID:        r.nextID,
		Title:     title,
		Genre:     genre,
		Director:  director,
		Year:      year,
		Watched:   watched,
		Rating:    rating,
		Notes:     notes,
		AddedDate: time.Now().Format("2006-01-02"),
	}
	r.movies = append(r.movies, movie)
	r.nextID++
	return movie, nil
}

func (r *MovieRandomizer) FindMovie(id int) *Movie {
	for i := range r.movies {
		if r.movies[i].ID == id {
			return &r.movies[i]
		}
	}
	return nil
}

func (r *MovieRandomizer) EditMovie(id int, updates map[string]interface{}) bool {
	movie := r.FindMovie(id)
	if movie == nil {
		return false
	}
	for key, value := range updates {
		switch key {
		case "title":
			if v, ok := value.(string); ok {
				movie.Title = v
			}
		case "genre":
			if v, ok := value.(string); ok {
				movie.Genre = v
			}
		case "director":
			if v, ok := value.(string); ok {
				movie.Director = v
			}
		case "year":
			if v, ok := value.(int); ok {
				movie.Year = v
			}
		case "watched":
			if v, ok := value.(bool); ok {
				movie.Watched = v
			}
		case "rating":
			if v, ok := value.(int); ok {
				movie.Rating = v
			}
		case "notes":
			if v, ok := value.(string); ok {
				movie.Notes = v
			}
		}
	}
	return true
}

func (r *MovieRandomizer) DeleteMovie(id int) bool {
	for i, m := range r.movies {
		if m.ID == id {
			r.movies = append(r.movies[:i], r.movies[i+1:]...)
			return true
		}
	}
	return false
}

func (r *MovieRandomizer) GetUnwatched() []Movie {
	var result []Movie
	for _, m := range r.movies {
		if !m.Watched {
			result = append(result, m)
		}
	}
	return result
}

func (r *MovieRandomizer) GetByGenre(genre string) []Movie {
	var result []Movie
	for _, m := range r.movies {
		if strings.EqualFold(m.Genre, genre) {
			result = append(result, m)
		}
	}
	return result
}

func (r *MovieRandomizer) RandomMovie(genre string) *Movie {
	pool := r.GetUnwatched()
	if genre != "" {
		pool = r.GetByGenre(genre)
	}
	if len(pool) == 0 {
		// Если нет подходящих, берём все (или по жанру)
		if genre != "" {
			pool = r.GetByGenre(genre)
		} else {
			pool = r.movies
		}
	}
	if len(pool) == 0 {
		return nil
	}
	return &pool[rand.Intn(len(pool))]
}

func (r *MovieRandomizer) GetStats() map[string]interface{} {
	total := len(r.movies)
	watched := 0
	for _, m := range r.movies {
		if m.Watched {
			watched++
		}
	}
	unwatched := total - watched
	var ratings []int
	for _, m := range r.movies {
		if m.Watched {
			ratings = append(ratings, m.Rating)
		}
	}
	avgRating := 0.0
	if len(ratings) > 0 {
		sum := 0
		for _, r := range ratings {
			sum += r
		}
		avgRating = float64(sum) / float64(len(ratings))
	}
	genres := make(map[string]int)
	for _, m := range r.movies {
		genres[m.Genre]++
	}
	return map[string]interface{}{
		"total":       total,
		"watched":     watched,
		"unwatched":   unwatched,
		"avg_rating":  avgRating,
		"genres":      genres,
	}
}

func (r *MovieRandomizer) SaveToFile(filename string) error {
	data := MoviesData{Movies: r.movies}
	jsonData, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(filename, jsonData, 0644)
}

func (r *MovieRandomizer) LoadFromFile(filename string) error {
	data, err := os.ReadFile(filename)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return err
	}
	var md MoviesData
	if err := json.Unmarshal(data, &md); err != nil {
		return err
	}
	r.movies = md.Movies
	for _, m := range r.movies {
		if m.ID >= r.nextID {
			r.nextID = m.ID + 1
		}
	}
	return nil
}

func (r *MovieRandomizer) ExportCSV(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()
	writer := csv.NewWriter(file)
	writer.Comma = ';'
	defer writer.Flush()
	headers := []string{"ID", "Название", "Жанр", "Режиссёр", "Год", "Просмотрен", "Оценка", "Заметки", "Дата добавления"}
	if err := writer.Write(headers); err != nil {
		return err
	}
	for _, m := range r.movies {
		watchedStr := "Нет"
		if m.Watched {
			watchedStr = "Да"
		}
		row := []string{
			strconv.Itoa(m.ID),
			m.Title,
			m.Genre,
			m.Director,
			strconv.Itoa(m.Year),
			watchedStr,
			strconv.Itoa(m.Rating),
			m.Notes,
			m.AddedDate,
		}
		if err := writer.Write(row); err != nil {
			return err
		}
	}
	return nil
}

func (r *MovieRandomizer) ImportCSV(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer file.Close()
	reader := csv.NewReader(file)
	reader.Comma = ';'
	reader.LazyQuotes = true
	records, err := reader.ReadAll()
	if err != nil {
		return err
	}
	if len(records) < 2 {
		return fmt.Errorf("файл пуст или нет данных")
	}
	for _, row := range records[1:] {
		if len(row) < 9 {
			continue
		}
		title := row[1]
		genre := row[2]
		director := row[3]
		year, _ := strconv.Atoi(row[4])
		watched := row[5] == "Да"
		rating, _ := strconv.Atoi(row[6])
		notes := row[7]
		_, err := r.AddMovie(title, genre, director, year, watched, rating, notes)
		if err != nil {
			fmt.Println("Ошибка импорта строки:", err)
		}
	}
	return nil
}

func readString(prompt string) string {
	fmt.Print(prompt)
	reader := bufio.NewReader(os.Stdin)
	input, _ := reader.ReadString('\n')
	return strings.TrimSpace(input)
}

func readInt(prompt string) int {
	for {
		input := readString(prompt)
		if val, err := strconv.Atoi(input); err == nil {
			return val
		}
		fmt.Println("Введите число.")
	}
}

func readBool(prompt string) bool {
	for {
		input := readString(prompt)
		if input == "1" {
			return true
		} else if input == "0" {
			return false
		}
		fmt.Println("Введите 1 или 0.")
	}
}

func printMovie(movie Movie) {
	status := "✅ Просмотрен"
	if !movie.Watched {
		status = "⏳ Не просмотрен"
	}
	fmt.Printf("#%d - %s (%d)\n", movie.ID, movie.Title, movie.Year)
	fmt.Printf("   Жанр: %s, Режиссёр: %s\n", movie.Genre, movie.Director)
	fmt.Printf("   %s, Оценка: %d/10\n", status, movie.Rating)
	if movie.Notes != "" {
		fmt.Printf("   Заметки: %s\n", movie.Notes)
	}
	fmt.Printf("   Добавлен: %s\n", movie.AddedDate)
}

func main() {
	randomizer := NewMovieRandomizer()
	if err := randomizer.LoadFromFile("movies_data.json"); err != nil {
		fmt.Println("Ошибка загрузки:", err)
	}

	for {
		fmt.Println("\n===== ГЕНЕРАТОР СЛУЧАЙНОГО ФИЛЬМА (Go) =====")
		fmt.Println("1. Добавить фильм")
		fmt.Println("2. Показать все фильмы")
		fmt.Println("3. Рекомендовать случайный фильм")
		fmt.Println("4. Рекомендовать по жанру")
		fmt.Println("5. Показать непросмотренные фильмы")
		fmt.Println("6. Отметить фильм как просмотренный")
		fmt.Println("7. Редактировать фильм")
		fmt.Println("8. Удалить фильм")
		fmt.Println("9. Показать статистику")
		fmt.Println("10. Сохранить в файл")
		fmt.Println("11. Загрузить из файла")
		fmt.Println("12. Экспорт в CSV")
		fmt.Println("13. Импорт из CSV")
		fmt.Println("0. Выход")
		choice := readString("Выберите действие: ")

		switch choice {
		case "0":
			return
		case "1":
			title := readString("Название: ")
			if title == "" {
				fmt.Println("Название не может быть пустым.")
				continue
			}
			genre := readString("Жанр: ")
			if genre == "" {
				fmt.Println("Жанр не может быть пустым.")
				continue
			}
			director := readString("Режиссёр (необязательно): ")
			year := readInt("Год выпуска: ")
			watched := readBool("Статус (1-просмотрен, 0-нет): ")
			rating := readInt("Оценка (1-10): ")
			notes := readString("Заметки (необязательно): ")
			movie, err := randomizer.AddMovie(title, genre, director, year, watched, rating, notes)
			if err != nil {
				fmt.Println("Ошибка:", err)
			} else {
				fmt.Printf("Фильм добавлен с ID %d\n", movie.ID)
			}
		case "2":
			if len(randomizer.movies) == 0 {
				fmt.Println("Нет фильмов.")
			} else {
				for _, m := range randomizer.movies {
					printMovie(m)
				}
			}
		case "3":
			movie := randomizer.RandomMovie("")
			if movie == nil {
				fmt.Println("Нет фильмов в коллекции.")
			} else {
				fmt.Println("\n🎬 Рекомендую посмотреть:")
				printMovie(*movie)
			}
		case "4":
			genre := readString("Введите жанр: ")
			if genre == "" {
				fmt.Println("Введите жанр.")
				continue
			}
			movie := randomizer.RandomMovie(genre)
			if movie == nil {
				fmt.Printf("Нет фильмов в жанре '%s'.\n", genre)
			} else {
				fmt.Printf("\n🎬 Рекомендую посмотреть в жанре %s:\n", genre)
				printMovie(*movie)
			}
		case "5":
			unwatched := randomizer.GetUnwatched()
			if len(unwatched) == 0 {
				fmt.Println("Нет непросмотренных фильмов.")
			} else {
				for _, m := range unwatched {
					printMovie(m)
				}
			}
		case "6":
			id := readInt("Введите ID фильма: ")
			movie := randomizer.FindMovie(id)
			if movie == nil {
				fmt.Println("Фильм не найден.")
				continue
			}
			if movie.Watched {
				fmt.Println("Фильм уже отмечен как просмотренный.")
			} else {
				movie.Watched = true
				fmt.Println("Фильм отмечен как просмотренный.")
			}
		case "7":
			id := readInt("Введите ID фильма для редактирования: ")
			movie := randomizer.FindMovie(id)
			if movie == nil {
				fmt.Println("Фильм не найден.")
				continue
			}
			fmt.Println("Оставьте поле пустым, чтобы не менять.")
			newTitle := readString(fmt.Sprintf("Название (%s): ", movie.Title))
			newGenre := readString(fmt.Sprintf("Жанр (%s): ", movie.Genre))
			newDirector := readString(fmt.Sprintf("Режиссёр (%s): ", movie.Director))
			newYear := readString(fmt.Sprintf("Год (%d): ", movie.Year))
			newWatched := readString(fmt.Sprintf("Статус (1-просмотрен, 0-нет) сейчас: %d: ", map[bool]int{true: 1, false: 0}[movie.Watched]))
			newRating := readString(fmt.Sprintf("Оценка (%d): ", movie.Rating))
			newNotes := readString(fmt.Sprintf("Заметки (%s): ", movie.Notes))
			updates := make(map[string]interface{})
			if newTitle != "" {
				updates["title"] = newTitle
			}
			if newGenre != "" {
				updates["genre"] = newGenre
			}
			if newDirector != "" {
				updates["director"] = newDirector
			}
			if newYear != "" {
				if y, err := strconv.Atoi(newYear); err == nil {
					updates["year"] = y
				} else {
					fmt.Println("Год должен быть числом, пропускаем.")
				}
			}
			if newWatched != "" {
				updates["watched"] = newWatched == "1"
			}
			if newRating != "" {
				if r, err := strconv.Atoi(newRating); err == nil {
					updates["rating"] = r
				} else {
					fmt.Println("Оценка должна быть числом, пропускаем.")
				}
			}
			if newNotes != "" {
				updates["notes"] = newNotes
			}
			if randomizer.EditMovie(id, updates) {
				fmt.Println("Фильм обновлён.")
			} else {
				fmt.Println("Ошибка обновления.")
			}
		case "8":
			id := readInt("Введите ID фильма для удаления: ")
			if randomizer.DeleteMovie(id) {
				fmt.Println("Фильм удалён.")
			} else {
				fmt.Println("Фильм не найден.")
			}
		case "9":
			stats := randomizer.GetStats()
			fmt.Println("\n=== СТАТИСТИКА ===")
			fmt.Printf("Всего фильмов: %d\n", stats["total"])
			fmt.Printf("Просмотрено: %d\n", stats["watched"])
			fmt.Printf("Не просмотрено: %d\n", stats["unwatched"])
			fmt.Printf("Средняя оценка (просмотренные): %.2f\n", stats["avg_rating"])
			fmt.Println("По жанрам:")
			genres := stats["genres"].(map[string]int)
			for g, c := range genres {
				fmt.Printf("  %s: %d\n", g, c)
			}
		case "10":
			if err := randomizer.SaveToFile("movies_data.json"); err != nil {
				fmt.Println("Ошибка сохранения:", err)
			} else {
				fmt.Println("Сохранено.")
			}
		case "11":
			if err := randomizer.LoadFromFile("movies_data.json"); err != nil {
				fmt.Println("Ошибка загрузки:", err)
			} else {
				fmt.Println("Загружено.")
			}
		case "12":
			if err := randomizer.ExportCSV("movies_export.csv"); err != nil {
				fmt.Println("Ошибка экспорта:", err)
			} else {
				fmt.Println("Экспортировано в movies_export.csv")
			}
		case "13":
			if err := randomizer.ImportCSV("movies_export.csv"); err != nil {
				fmt.Println("Ошибка импорта:", err)
			} else {
				fmt.Println("Импортировано из movies_export.csv")
			}
		default:
			fmt.Println("Неизвестная команда.")
		}
	}
}

File: movie_randomizer.js
// movie_randomizer.js
const fs = require('fs').promises;
const readline = require('readline');
const { parse } = require('csv-parse');
const { createObjectCsvWriter } = require('csv-writer');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const question = (prompt) => new Promise(resolve => rl.question(prompt, resolve));

class Movie {
    constructor(id, title, genre, director, year, watched, rating, notes, addedDate) {
        this.id = id;
        this.title = title;
        this.genre = genre;
        this.director = director || 'Неизвестен';
        this.year = year;
        this.watched = watched;
        this.rating = rating;
        this.notes = notes || '';
        this.addedDate = addedDate || new Date().toISOString().slice(0, 10);
    }
}

class MovieRandomizer {
    constructor() {
        this.movies = [];
        this.nextId = 1;
    }

    addMovie(title, genre, director, year, watched, rating, notes) {
        if (rating < 1 || rating > 10) throw new Error('Оценка должна быть от 1 до 10');
        const currentYear = new Date().getFullYear();
        if (year < 1900 || year > currentYear) throw new Error(`Год должен быть от 1900 до ${currentYear}`);
        if (!title.trim() || !genre.trim()) throw new Error('Название и жанр не могут быть пустыми');
        if (!director.trim()) director = 'Неизвестен';
        const movie = new Movie(this.nextId, title, genre, director, year, watched, rating, notes);
        this.movies.push(movie);
        this.nextId++;
        return movie;
    }

    findMovie(id) {
        return this.movies.find(m => m.id === id);
    }

    editMovie(id, updates) {
        const movie = this.findMovie(id);
        if (!movie) return false;
        Object.assign(movie, updates);
        return true;
    }

    deleteMovie(id) {
        const index = this.movies.findIndex(m => m.id === id);
        if (index === -1) return false;
        this.movies.splice(index, 1);
        return true;
    }

    getUnwatched() {
        return this.movies.filter(m => !m.watched);
    }

    getByGenre(genre) {
        return this.movies.filter(m => m.genre.toLowerCase() === genre.toLowerCase());
    }

    randomMovie(genre) {
        let pool = genre ? this.getByGenre(genre) : this.getUnwatched();
        if (pool.length === 0) {
            // если нет, берём все (или по жанру)
            pool = genre ? this.getByGenre(genre) : this.movies;
        }
        if (pool.length === 0) return null;
        return pool[Math.floor(Math.random() * pool.length)];
    }

    getStats() {
        const total = this.movies.length;
        const watchedCount = this.movies.filter(m => m.watched).length;
        const unwatched = total - watchedCount;
        const ratings = this.movies.filter(m => m.watched).map(m => m.rating);
        const avgRating = ratings.length ? ratings.reduce((a, b) => a + b, 0) / ratings.length : 0;
        const genres = {};
        this.movies.forEach(m => {
            genres[m.genre] = (genres[m.genre] || 0) + 1;
        });
        return { total, watched: watchedCount, unwatched, avgRating, genres };
    }

    async saveToFile(filename = 'movies_data.json') {
        const data = { movies: this.movies };
        await fs.writeFile(filename, JSON.stringify(data, null, 2));
    }

    async loadFromFile(filename = 'movies_data.json') {
        try {
            const data = await fs.readFile(filename, 'utf8');
            const parsed = JSON.parse(data);
            this.movies = parsed.movies.map(m => Object.assign(new Movie(0), m));
            this.nextId = this.movies.reduce((max, m) => Math.max(max, m.id), 0) + 1;
        } catch (err) {
            if (err.code !== 'ENOENT') throw err;
        }
    }

    async exportCSV(filename = 'movies_export.csv') {
        const records = this.movies.map(m => ({
            ID: m.id,
            Название: m.title,
            Жанр: m.genre,
            Режиссёр: m.director,
            Год: m.year,
            Просмотрен: m.watched ? 'Да' : 'Нет',
            Оценка: m.rating,
            Заметки: m.notes,
            'Дата добавления': m.addedDate
        }));
        const csvWriter = createObjectCsvWriter({
            path: filename,
            header: [
                { id: 'ID', title: 'ID' },
                { id: 'Название', title: 'Название' },
                { id: 'Жанр', title: 'Жанр' },
                { id: 'Режиссёр', title: 'Режиссёр' },
                { id: 'Год', title: 'Год' },
                { id: 'Просмотрен', title: 'Просмотрен' },
                { id: 'Оценка', title: 'Оценка' },
                { id: 'Заметки', title: 'Заметки' },
                { id: 'Дата добавления', title: 'Дата добавления' }
            ],
            fieldDelimiter: ';'
        });
        await csvWriter.writeRecords(records);
    }

    async importCSV(filename = 'movies_export.csv') {
        const fileContent = await fs.readFile(filename, 'utf8');
        return new Promise((resolve, reject) => {
            parse(fileContent, { columns: true, delimiter: ';' }, (err, records) => {
                if (err) reject(err);
                for (const row of records) {
                    try {
                        this.addMovie(
                            row['Название'],
                            row['Жанр'],
                            row['Режиссёр'],
                            parseInt(row['Год']),
                            row['Просмотрен'] === 'Да',
                            parseInt(row['Оценка']),
                            row['Заметки']
                        );
                    } catch (e) {
                        console.log('Ошибка импорта строки:', e.message);
                    }
                }
                resolve();
            });
        });
    }
}

function printMovie(movie) {
    const status = movie.watched ? '✅ Просмотрен' : '⏳ Не просмотрен';
    console.log(`#${movie.id} - ${movie.title} (${movie.year})`);
    console.log(`   Жанр: ${movie.genre}, Режиссёр: ${movie.director}`);
    console.log(`   ${status}, Оценка: ${movie.rating}/10`);
    if (movie.notes) console.log(`   Заметки: ${movie.notes}`);
    console.log(`   Добавлен: ${movie.addedDate}`);
}

async function main() {
    const randomizer = new MovieRandomizer();
    await randomizer.loadFromFile();

    while (true) {
        console.log('\n===== ГЕНЕРАТОР СЛУЧАЙНОГО ФИЛЬМА (JavaScript) =====');
        console.log('1. Добавить фильм');
        console.log('2. Показать все фильмы');
        console.log('3. Рекомендовать случайный фильм');
        console.log('4. Рекомендовать по жанру');
        console.log('5. Показать непросмотренные фильмы');
        console.log('6. Отметить фильм как просмотренный');
        console.log('7. Редактировать фильм');
        console.log('8. Удалить фильм');
        console.log('9. Показать статистику');
        console.log('10. Сохранить в файл');
        console.log('11. Загрузить из файла');
        console.log('12. Экспорт в CSV');
        console.log('13. Импорт из CSV');
        console.log('0. Выход');
        const choice = await question('Выберите действие: ');

        if (choice === '0') break;

        switch (choice) {
            case '1': {
                const title = await question('Название: ');
                if (!title.trim()) { console.log('Название не может быть пустым.'); continue; }
                const genre = await question('Жанр: ');
                if (!genre.trim()) { console.log('Жанр не может быть пустым.'); continue; }
                const director = await question('Режиссёр (необязательно): ');
                const year = parseInt(await question('Год выпуска: '));
                const watched = (await question('Статус (1-просмотрен, 0-нет): ')) === '1';
                const rating = parseInt(await question('Оценка (1-10): '));
                const notes = await question('Заметки (необязательно): ');
                try {
                    const movie = randomizer.addMovie(title, genre, director, year, watched, rating, notes);
                    console.log(`Фильм добавлен с ID ${movie.id}`);
                } catch (err) {
                    console.log('Ошибка:', err.message);
                }
                break;
            }
            case '2':
                if (randomizer.movies.length === 0) console.log('Нет фильмов.');
                else randomizer.movies.forEach(printMovie);
                break;
            case '3': {
                const movie = randomizer.randomMovie();
                if (!movie) console.log('Нет фильмов в коллекции.');
                else { console.log('\n🎬 Рекомендую посмотреть:'); printMovie(movie); }
                break;
            }
            case '4': {
                const genre = await question('Введите жанр: ');
                if (!genre.trim()) { console.log('Введите жанр.'); continue; }
                const movie = randomizer.randomMovie(genre);
                if (!movie) console.log(`Нет фильмов в жанре '${genre}'.`);
                else { console.log(`\n🎬 Рекомендую посмотреть в жанре ${genre}:`); printMovie(movie); }
                break;
            }
            case '5': {
                const unwatched = randomizer.getUnwatched();
                if (unwatched.length === 0) console.log('Нет непросмотренных фильмов.');
                else unwatched.forEach(printMovie);
                break;
            }
            case '6': {
                const id = parseInt(await question('Введите ID фильма: '));
                const movie = randomizer.findMovie(id);
                if (!movie) { console.log('Фильм не найден.'); continue; }
                if (movie.watched) console.log('Фильм уже отмечен как просмотренный.');
                else { randomizer.editMovie(id, { watched: true }); console.log('Фильм отмечен как просмотренный.'); }
                break;
            }
            case '7': {
                const id = parseInt(await question('Введите ID фильма для редактирования: '));
                const movie = randomizer.findMovie(id);
                if (!movie) { console.log('Фильм не найден.'); continue; }
                console.log('Оставьте поле пустым, чтобы не менять.');
                const newTitle = await question(`Название (${movie.title}): `);
                const newGenre = await question(`Жанр (${movie.genre}): `);
                const newDirector = await question(`Режиссёр (${movie.director}): `);
                const newYear = await question(`Год (${movie.year}): `);
                const newWatched = await question(`Статус (1-просмотрен, 0-нет) сейчас: ${movie.watched ? '1' : '0'}: `);
                const newRating = await question(`Оценка (${movie.rating}): `);
                const newNotes = await question(`Заметки (${movie.notes}): `);
                const updates = {};
                if (newTitle.trim()) updates.title = newTitle;
                if (newGenre.trim()) updates.genre = newGenre;
                if (newDirector.trim()) updates.director = newDirector;
                if (newYear.trim()) {
                    const y = parseInt(newYear);
                    if (!isNaN(y)) updates.year = y;
                    else console.log('Год должен быть числом, пропускаем.');
                }
                if (newWatched.trim()) updates.watched = newWatched === '1';
                if (newRating.trim()) {
                    const r = parseInt(newRating);
                    if (!isNaN(r)) updates.rating = r;
                    else console.log('Оценка должна быть числом, пропускаем.');
                }
                if (newNotes.trim()) updates.notes = newNotes;
                if (randomizer.editMovie(id, updates)) console.log('Фильм обновлён.');
                else console.log('Ошибка обновления.');
                break;
            }
            case '8': {
                const id = parseInt(await question('Введите ID фильма для удаления: '));
                if (randomizer.deleteMovie(id)) console.log('Фильм удалён.');
                else console.log('Фильм не найден.');
                break;
            }
            case '9': {
                const stats = randomizer.getStats();
                console.log('\n=== СТАТИСТИКА ===');
                console.log(`Всего фильмов: ${stats.total}`);
                console.log(`Просмотрено: ${stats.watched}`);
                console.log(`Не просмотрено: ${stats.unwatched}`);
                console.log(`Средняя оценка (просмотренные): ${stats.avgRating.toFixed(2)}`);
                console.log('По жанрам:');
                for (const [g, c] of Object.entries(stats.genres)) {
                    console.log(`  ${g}: ${c}`);
                }
                break;
            }
            case '10':
                try {
                    await randomizer.saveToFile();
                    console.log('Сохранено.');
                } catch (err) {
                    console.log('Ошибка сохранения:', err.message);
                }
                break;
            case '11':
                try {
                    await randomizer.loadFromFile();
                    console.log('Загружено.');
                } catch (err) {
                    console.log('Ошибка загрузки:', err.message);
                }
                break;
            case '12':
                try {
                    await randomizer.exportCSV();
                    console.log('Экспортировано в movies_export.csv');
                } catch (err) {
                    console.log('Ошибка экспорта:', err.message);
                }
                break;
            case '13':
                try {
                    await randomizer.importCSV();
                    console.log('Импортировано из movies_export.csv');
                } catch (err) {
                    console.log('Ошибка импорта:', err.message);
                }
                break;
            default:
                console.log('Неизвестная команда.');
        }
    }
    rl.close();
}

main().catch(console.error);
